src/qt/nametablemodel.cpp

```diff
@@ -0,0 +1,566 @@
+        }
+
+        if(pendingNames.isArray())
+        {
+            for (const auto& v : pendingNames.getValues())
+            {
+                std::string name = find_value ( v, "name").get_str();
+                std::string data = find_value ( v, "value").get_str();
+                vNamesO[name] = NameTableEntry(name, data, NameTableEntry::NAME_UNCONFIRMED);
+                LogPrintf("found pending name: name=%s\n", name.c_str());
+            }
+        }
+
+        // confirmed names (name_list)
+        JSONRPCRequest nameListRequest;
+        nameListRequest.strMethod = "name_list";
+        nameListRequest.params = NullUniValue;
+        nameListRequest.fHelp = false;
+        UniValue confirmedNames;
+
+        try {
+            confirmedNames = tableRPC.execute(nameListRequest);
+        } catch (const UniValue& e) {
+            UniValue message = find_value( e, "message");
+            LogPrintf ("name_list lookup error: %s\n", message.get_str().c_str());
+        }
+
+        // will be an object if name_list command isn't available/other error
+        if(confirmedNames.isArray())
+        {
+            for (const auto& v : confirmedNames.getValues())
+            {
+                //const UniValue& v = confirmedNames[idx];
+                std::string name = find_value ( v, "name").get_str();
+                std::string data = find_value ( v, "value").get_str();
+                int height = find_value ( v, "height").get_int();
+                vNamesO[name] = NameTableEntry(name, data, height);
+                LogPrintf("found confirmed name: name=%s height=%i\n", name.c_str(), height);
+            }
+        }
+
+        // Add existing names
+        for (const auto& item : vNamesO)
+            cachedNameTable.append(item.second);
+
+        // Add pending names (name_new)
+        LOCK(wallet->cs_wallet);
+        for (const auto& item : wallet->namePendingMap)
+        {
+            CNamePendingData npd = CNamePendingData(item.second);
+            cachedNameTable.append(
+                NameTableEntry(item.first,
+                               npd.getData(),
+                               NameTableEntry::NAME_NEW));
+        }
+
+        // qLowerBound() and qUpperBound() require our cachedNameTable list to be sorted in asc order
+        qSort(cachedNameTable.begin(), cachedNameTable.end(), NameTableEntryLessThan());
+    }
+
+    bool findInModel(const QString &name, int *lowerIndex=nullptr, int *upperIndex=nullptr)
+    {
+        // Find name in model
+        QList<NameTableEntry>::iterator lower = qLowerBound(
+            cachedNameTable.begin(), cachedNameTable.end(), name, NameTableEntryLessThan());
+        QList<NameTableEntry>::iterator upper = qUpperBound(
+            cachedNameTable.begin(), cachedNameTable.end(), name, NameTableEntryLessThan());
+        if(lowerIndex)
+            *lowerIndex = (lower - cachedNameTable.begin());
```

Comment:
Are `QList` iterators random access (so that you can take the iterator difference efficiently)?  I don't know what a `QList` is, but the name "list" for me implies a linked list (like `std::list`), in which case they are not.

Reply:
`qUpperBound` and `qLowerBound` takes `RandomAccessIterators` as params, so I assume they're fairly performant. I'll need to refactor this stuff eventually for future Qt versions, though. Can this stay for now or is it a blocker? I'm planning on doing a lot more refactoring of this class in general in the future.

Reply:
No, this is definitely not a blocker - I was just curious.  (And I may even be wrong, possibly `QList` *does* have random-access iterators in which case my remark is entirely moot anyway.)  So if you want to refactor this anyway, feel free to keep as it is for now.